packages/react/src/ReactElementValidator.js

```diff
@@ -177,57 +180,93 @@ function validateChildKeys(node, parentType) {
   }
 }
 
-/**
- * Given an element, validate that its props follow the propTypes definition,
- * provided by the type.
- *
- * @param {ReactElement} element
- */
-function validatePropTypes(element) {
-  const type = element.type;
-  let name, propTypes;
+function validatePropTypesRecursively(type, props) {
+  if (type === null || type === undefined || typeof type === 'string') {
+    return;
+  }
+  let propTypes;
+  let innerType;
   if (typeof type === 'function') {
-    // Class or function component
-    name = type.displayName || type.name;
-    propTypes = type.propTypes;
-  } else if (
-    typeof type === 'object' &&
-    type !== null &&
-    type.$$typeof === REACT_FORWARD_REF_TYPE
-  ) {
-    // ForwardRef
-    const functionName = type.render.displayName || type.render.name || '';
-    name =
-      type.displayName ||
-      (functionName !== '' ? `ForwardRef(${functionName})` : 'ForwardRef');
     propTypes = type.propTypes;
-  } else {
-    return;
+  } else if (typeof type === 'object') {
+    switch (type.$$typeof) {
+      case REACT_FORWARD_REF_TYPE: {
+        propTypes = type.propTypes;
+        break;
+      }
+      case REACT_LAZY_TYPE: {
+        propTypes = type.propTypes;
+        innerType = refineResolvedLazyComponent(type);
+        break;
+      }
+      case REACT_MEMO_TYPE: {
+        propTypes = type.propTypes;
+        innerType = type.type;
+        break;
+      }
+    }
   }
+  // Common case.
   if (propTypes) {
-    setCurrentlyValidatingElement(element);
+    const name = getComponentName(type);
     checkPropTypes(
       propTypes,
-      element.props,
+      props,
       'prop',
       name,
       ReactDebugCurrentFrame.getStackAddendum,
     );
-    setCurrentlyValidatingElement(null);
-  } else if (type.PropTypes !== undefined && !propTypesMisspellWarningShown) {
-    propTypesMisspellWarningShown = true;
-    warningWithoutStack(
-      false,
-      'Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?',
-      name || 'Unknown',
-    );
   }
-  if (typeof type.getDefaultProps === 'function') {
-    warningWithoutStack(
-      type.getDefaultProps.isReactClassApproved,
-      'getDefaultProps is only used on classic React.createClass ' +
-        'definitions. Use a static property named `defaultProps` instead.',
-    );
+  // A type may have an inner type (e.g. React.memo or React.lazy).
+  // That's what the recursive case is for.
+  if (innerType) {
+    // Inner type may have its own defaultProps
+    let innerProps = props;
+    if (innerType && innerType.defaultProps) {
+      innerProps = {...props};
```

Comment:
feels like a bit of a waste we need to do this and throw it away, I guess

Reply:
Not the common case though? I'm assuming `defaultProps` is not super common by itself, and especially when wrapped.